### ISSUE TITLE
Seems that the abs activation is redundant?

### ISSUE THREAD

**Issue by yzslab** [humansensinglab/taming-3dgs#13]
Hi, I found that the opacities will not be optimized after replacing the activation function with the abs [here](https://github.com/humansensinglab/taming-3dgs/blob/7553b073db2d778f50183e5164e802d5e5cb1af9/scene/gaussian_model.py#L54-L58), due to the absence of the `replace_tensor_to_optimizer` operation.
Since the range of the Sigmoid is $$(0, 1)$$ and the opacities are fixed, they will never fall below 0. Therefore, it seems that the abs activation is redundant, as simply returning the same tensor suffices.
I tested removing the abs by the modification below and it produced the same metrics:
```diff
diff --git a/scene/gaussian_model.py b/scene/gaussian_model.py
index 8c65c4c..0e46ba8 100644
--- a/scene/gaussian_model.py
+++ b/scene/gaussian_model.py
@@ -46,14 +46,14 @@ class GaussianModel:
             self.inverse_opacity_activation = inverse_sigmoid
         else:
             print("Absolute rendering mode")
-            self.opacity_activation = torch.abs
+            self.opacity_activation = identity_gate
             self.inverse_opacity_activation = identity_gate
 
         self.rotation_activation = torch.nn.functional.normalize
 
     def modify_functions(self):
         old_opacities = self.get_opacity.clone()
-        self.opacity_activation = torch.abs
+        self.opacity_activation = identity_gate
         self.inverse_opacity_activation = identity_gate
         self._opacity = self.opacity_activation(old_opacities)
```
Is this expected?

**Comment by saswat0** [comment]
@yzslab You are correct. The abs activation is kept just as a failsafe and we don't expect any performance improvements with it

**Comment by yzslab** [comment]
> @yzslab You are correct. The abs activation is kept just as a failsafe and we don't expect any performance improvements with it

Hi, does the `modify_functions` correspond to the Section 3.4 of your paper?

> Therefore, we convert the regular, capped Gaussian primitives to high-opacity Gaussians after reaching the midpoint of our training (15K iterations). This involves replacing the opacity activation with abs and clamping blending weights to 1 from above during rendering. As shown by our ablation, this change positively impacts quality metrics, particularly PSNR.

Could you please confirm if this conclusion is still valid?

**Comment by saswat0** [comment]
@yzslab Apologies for the late response.

Thanks for pointing out this issue. We've patched the bug with [5aef77d](https://github.com/humansensinglab/taming-3dgs/commit/5aef77d0624aad195a4e92fa69bd707381cb8e45)

**Comment by yzslab** [comment]
> @yzslab Apologies for the late response.
> 
> Thanks for pointing out this issue. We've patched the bug with [d1fced9](https://github.com/humansensinglab/taming-3dgs/commit/d1fced9bfe747226886ba4b272d9714a4240dfec)

Thanks, but have you tested it? Seems that such a simple solution can not produce the expected results.

**Comment by yzslab** [comment]
The unexpected result with the high-opacity Gaussians is related to the early stopping and can be simply fixed in this way:
```diff
diff --git a/submodules/diff-gaussian-rasterization/cuda_rasterizer/forward.cu b/submodules/diff-gaussian-rasterization/cuda_rasterizer/forward.cu
index c541381..5720d7a 100644
--- a/submodules/diff-gaussian-rasterization/cuda_rasterizer/forward.cu
+++ b/submodules/diff-gaussian-rasterization/cuda_rasterizer/forward.cu
@@ -388,12 +388,12 @@ renderCUDA(
                        float alpha = min(0.99f, con_o.w * exp(power));
                        if (alpha < 1.0f / 255.0f)
                                continue;
-                       float test_T = T * (1 - alpha);
-                       if (test_T < 0.0001f)
+                       if (T < 0.0001f)
                        {
                                done = true;
                                continue;
                        }
+                       float test_T = T * (1 - alpha);
 
                        // Eq. (3) from 3D Gaussian splatting paper.
                        for (int ch = 0; ch < CHANNELS; ch++)
```

But the metrics are still lower than normal opacity.
